src/node_crypto.cc

```diff
@@ -1391,6 +1399,188 @@ void SecureContext::GetCertificate(const FunctionCallbackInfo<Value>& args) {
 }
 
 
+#ifdef OPENSSL_PSK_SUPPORT
+
+void SecureContext::SetPskIdentity(const FunctionCallbackInfo<Value>& args) {
+  SecureContext* wrap = Unwrap<SecureContext>(args.Holder());
+
+  if (!args[0]->IsString()) {
+    return wrap->env()->ThrowTypeError("Argument must be a string");
+  }
+
+  String::Utf8Value identity(args.GetIsolate(), args[0]);
+  if (!SSL_CTX_use_psk_identity_hint(wrap->ctx_, *identity)) {
+    return wrap->env()->ThrowError("Failed to set PSK identity hint");
+  }
+}
+
+void SecureContext::EnablePskCallback(
+    const FunctionCallbackInfo<Value>& args) {
+  SecureContext* wrap = Unwrap<SecureContext>(args.Holder());
+
+  SSL_CTX_set_psk_server_callback(wrap->ctx_,
+                                  SecureContext::PskServerCallback);
+  SSL_CTX_set_psk_client_callback(wrap->ctx_,
+                                  SecureContext::PskClientCallback);
+}
+
+unsigned int SecureContext::PskServerCallback(SSL *ssl,
+                                              const char *identity,
+                                              unsigned char *psk,
+                                              unsigned int max_psk_len) {
+  SecureContext* sc = static_cast<SecureContext*>(
+      SSL_CTX_get_app_data(SSL_get_SSL_CTX(ssl)));
+
+  Environment* env = sc->env();
+  Isolate* isolate = env->isolate();
+  HandleScope scope(isolate);
+
+  MaybeLocal<String> hint_str = String::NewFromUtf8(
+    isolate,
+    identity,
+    v8::NewStringType::kNormal,
+    strlen(identity));
+
+  if (hint_str.IsEmpty()) {
+    return 0;
+  }
+
+  Local<Value> argv[] = {
+    hint_str.ToLocalChecked(),
+    Integer::NewFromUnsigned(isolate, max_psk_len),
+    Integer::NewFromUnsigned(isolate, 0)
+  };
+
+  Local<Value> value = sc->object()->Get(env->onpskexchange_string());
+
+  MaybeLocal<Value> maybe_ret;
+  if (value->IsFunction()) {
+    Local<Function> func = value.As<Function>();
+    maybe_ret = func->Call(env->context(), sc->object(), arraysize(argv), argv);
+  } else {
+    return 0;
+  }
+
+  if (maybe_ret.IsEmpty()) {
+    return 0;
+  }
+
+  Local<Value> ret = maybe_ret.ToLocalChecked();
+
+  // The result is expected to be an object. If it isn't, then return 0,
+  // indicating the identity wasn't found.
+  if (!ret->IsObject()) {
+    return 0;
+  }
+  Local<Object> obj = ret.As<Object>();
+
+  Local<Value> psk_val = obj->Get(env->context(),
+                                  env->psk_string()).ToLocalChecked();
+  CHECK(psk_val->IsString());
+  Local<String> psk_str = psk_val.As<String>();
+
+  Local<String> enc = String::Empty(isolate);
+  StringBytes::InlineDecoder decoder;
+  if (!decoder.Decode(env, psk_str, enc, HEX))
+    return 0;
+
+  char* psk_buf = decoder.out();
+  size_t psk_len = decoder.size();
+
+  CHECK_LE(psk_len, max_psk_len);
```

Comment:
Likewise.  Perhaps these should throw JS exceptions as well.

Reply:
Updated these, as I mentioned above, I was also doing JS validation and exception throwing here as well 😁